src/charm.py

```diff
@@ -301,16 +302,60 @@ def _prometheus_config(self):
 
         return yaml.dump(scrape_config)
 
+    def _get_image_tag(self, api_tags=None) -> str:
+        """Get most recent patch of major and minor Prometheus version
+        """
+        # 1. get valid tags from URL (or directly from input)
+        # 2. clean tags data structure and filter out release candidates "rc"
+        # 3. filter out tags that don't match our major/minor specification
+        config = self.model.config
+        if not api_tags:
+            api_tags_resp = requests.get(config['image-tags-url'])
+            if not api_tags_resp.ok:
+                logger.error('No Prometheus image tag can be selected. Cannot access '
+                             '{} (defined in config.yaml).'.format(config['image-tags-url']))
+                self.unit.status = BlockedStatus('Cannot get container image path')
+                return ''
+            api_tags = json.loads(api_tags_resp.text)
+
+        major = config['prometheus-image-major-version']
+        minor = config['prometheus-image-minor-version']
+        tags = [tag['name'].split('.') for tag in api_tags if 'rc' not in tag['name']]
+        tags = [tag for tag in tags
+                if len(tag) >= 2 and tag[0].replace('v', '') == major and tag[1] == minor]
+
+        # if there is nothing in tags, we haven't found a valid set of tags
+        if not tags:
+            logger.error('Prometheus version cannot be verified: v{0}.{1}'.format(major, minor))
+            self.unit.status = BlockedStatus('Cannot verify prom v{0}.{1}'.format(major, minor))
+            return ''
+
+        # select the latest patch version for v{major}.{minor}
+        patched_tags = [tag for tag in tags if len(tag) == 3]
+        unpatched_tags = [tag for tag in tags if len(tag) == 2]
+        sorted_tags = sorted(patched_tags, key=lambda tag: tag[2], reverse=True)
```

Comment:
Doesn't this sort based on string and not based on int?
So it would do weird things with 2.2.1 vs 2.2.10 vs 2.2.2
And we also have to watch out for tags like 7.2.3-ubuntu, right?
Though I guess that is Grafana tag structure:
https://hub.docker.com/r/grafana/grafana/tags

not Prometheus.
But I do worry about str vs int sorting. (prometheus definitely has 2.20, but I don't know if they hit 2.20.10)

Reply:
It does sort based on string, but since we do `tag.split('.')` the only string comparisons (at least for Prometheus) will be stuff like '2' vs '20' where the string '20' would be larger than '2'.

Like I mentioned in the above comment, I think this logic will only work for Prometheus tags (and repos with similar tag structures). Right now, I'm not sure how we can generalize this. I just wanted to make sure we had something working.

Reply:
I think you're missing how this works. If you have 2.20.1, 2.20.2, ... 2.20.9, 2.20.10, ,2.20.11
Then string sorting will put 2.20.2 "higher" than 2.20.11

Reply:
Brain fart. Let me correct this.

Reply:
Might be worth leveraging https://pypi.org/project/semver/ for this logic.

Reply:
Oh that's super helpful @jlounder . I wasn't aware of that.